### ISSUE TITLE
time-cost error (torchserve:0.5.2-cpu VS torchserve:0.5.2-gpu)

### ISSUE THREAD

**Issue by Flyingdog-Huang** [pytorch/serve#1525]
Please have a look at [FAQ's](../../docs/FAQs.md) and [Troubleshooting guide](../../docs/Troubleshooting.md), your query may be already addressed.

Your issue may already be reported!
Please search on the [issue tracker](https://github.com/pytorch/serve/issues) before creating one.

## Context

when I use docker torchserve:0.5.2-cpu, model  inference(with CPU) cost time:27s, and  post process(with CPU)  cost time:7s;
![cpu](https://user-images.githubusercontent.com/43697342/159434334-4f09b071-d493-42ed-990f-9f2e0a2db093.png)

when I use docker torchserve:0.5.2-gpu, model  inference(with GPU)  cost time:1.7s, and  post process(with CPU)  cost time:55s.
![gpu](https://user-images.githubusercontent.com/43697342/159434373-82ef5ae6-8367-4f23-acf9-8b6e26c61abb.png)

why are there so much differences between docker -cpu and docker -gpu when I post process with same CPU?

* torchserve version: docker  (torchserve:0.5.2-cpu VS torchserve:0.5.2-gpu)
* torch-model-archiver version:
* torch version:
* torchvision version [if any]:
* torchtext version [if any]:
* torchaudio version [if any]:
* java version:
* Operating System and version: Linux

## Your Environment

* Installed using source? [yes
* Are you planning to deploy it using docker container? [yes/
* Is it a CPU or GPU environment?: cpu and gpu
* Using a default/custom handler? [If possible upload/share custom handler/model]:custom handler(BaseHandler)
* What kind of model is it e.g. vision, text, audio?:vision-Unet
* Are you planning to use local models from model-store or public url being used e.g. from S3 bucket etc.?
  [If public url then provide link.]: local model
* Provide config.properties, logs [ts.log] and parameters used for model registration/update APIs:
* log-cpu
![cpu](https://user-images.githubusercontent.com/43697342/159432074-978eb332-7827-4260-919e-d0151cf762f8.png)
log-gpu
![gpu](https://user-images.githubusercontent.com/43697342/159432185-9a1b91cb-8f7a-48c1-93ad-1f047b1c79ac.png)

* 
* Link to your project [if any]:

## Possible Solution

can docker -cpu and docker-gpu use CPU with the same way?

## Steps to Reproduce

1.
2.
...

**Comment by msaroufim** [comment]
Do you mind sharing what your `postprocess()` function looks like, if those operations are all happening on CPU you could be seeing lengthy data transfer costs from CPU to GPU in which case. There are ways of doing more directly on GPU but would be helpful to see the example

Could you also share the command you're using to start docker?

**Comment by Flyingdog-Huang** [comment]
> Do you mind sharing what your `postprocess()` function looks like, if those operations are all happening on CPU you could be seeing lengthy data transfer costs from CPU to GPU in which case. There are ways of doing more directly on GPU but would be helpful to see the example
> 
> Could you also share the command you're using to start docker?

Thank you for your reply. @msaroufim 

1, this is my postprocess()` function:
    def postprocess(self, pre_data, inf_data):
        logger.info("postprocess starting")
        device_name="cpu"
        self.device = torch.device(device_name)
        wall_vectors=[]
        for image,probs in zip(pre_data,inf_data):
            # process pridiction data - (h,w,c)
            time_process=time.perf_counter()
            probs = probs.to(device=self.device)
            logger.info("data device : {}".format(probs.device))
            time_gpu2cpu=time.perf_counter()
            logger.info("gpu2cpu time : {}".format(time_gpu2cpu-time_process))
            probs = F.softmax(probs, dim=1)[0]
            tf = T.Compose([T.ToPILImage(),
                            T.Resize((image.shape[0], image.shape[1])),
                            T.ToTensor()])
            full_mask = tf(probs).squeeze()
            full_mask = F.one_hot(full_mask.argmax(dim=0), n_classes).permute(2, 0, 1).numpy()
            result_img = np.uint8(full_mask[1]*255)
            img_pre=np.transpose(np.array([result_img,result_img,result_img]),(1,2,0))
            time_tensor2img=time.perf_counter()
            logger.info("tensor2img time : {}".format(time_tensor2img-time_gpu2cpu))
            
            # verctorizaton
            wall_vector_i=wall_vector(image,img_pre)
            wall_vectors.append(wall_vector_i)
            time_vector=time.perf_counter()
            logger.info("verctorizaton time : {}".format(time_vector-time_tensor2img))
        return wall_vectors

### the main part of post process  is wall_vector() , in this funtion , the types of input data(image,img_pre) are also numpy.array,  and the libs are only numpy and opencv.

2, about data transfer costs :
in this case, I just process only one img data, the size is (3500*2500*3),  and also transfer this size data from GPU to CPU ,  the cost time is 1s(I have tested).

3,  the command you're using to start docker:
CPU: docker run --rm -it  ... pytorch/torchserve:0.5.2-cpu
GPU: docker run --rm -it --gpus all   ...  pytorch/torchserve:0.5.2-gpu


BTW, I use the same codes in cpu and gpu docker, I set self.device as 'torch.device("cuda:" + str(properties.get("gpu_id")) if torch.cuda.is_available() else "cpu")'  in initialize(),  and I change it in postprocess(), that you can see the codes above.

thank you so much again.

**Comment by Flyingdog-Huang** [comment]
@msaroufim 
Thank you for your reply. 
And I find the error that I forgot load the model weight file .pth for the inference model.
I am so sorry for my carelessness, and I will close this issue .
Thank you again for your kind and patient, hope you have a nice day.
